Discuss ACCUMULO-3079: collapsing the iterator stack to improve performance

milleruntime opened the pull request:
Applied the changes from the patch submitted by Adam Fuchs on [ACCUMULO-3079](https://issues.apache.org/jira/browse/ACCUMULO-3079) to master. This PR is to bring these changes back into discussion.  Created JMH benchmark tests [here](https://github.com/milleruntime/jmh-test/blob/master/src/main/java/org/sample/MyBenchmark.java).  I tried to isolate the tests to only the system iterators acting on data in memory.  

I have attached the results from the benchmark tests. 
[jmh-accumulo-benchmark-results.txt](https://github.com/apache/accumulo/files/906333/jmh-accumulo-benchmark-results.txt)

The JMH tests can be built and run from [here](https://github.com/milleruntime/jmh-test). After cloning the repo, simply run:
`mvn clean install`
`java -jar target/benchmarks.jar -i 100 -f 1`

joshelser commented:
> VM version: JDK 1.8.0_121, VM 25.121-b13
> VM invoker: /Library/Java/JavaVirtualMachines/jdk1.8.0_121.jdk/Contents/Home/jre/bin/java

We should definitely run these on Linux.

milleruntime commented:
Here are updated results with @keith-turner 's additional improvements: 
[testresults.txt](https://github.com/apache/accumulo/files/911586/testresults.txt)

joshelser commented:
>  Created JMH benchmark tests here.

Also, I would be super in-favor with including this in Accumulo. I think there are some licensing things that we would have to figure out (JMH is GPL IIRC), but I think this would be worthwhile as a benchmark is no good if no one else ever runs it :)

scubafuchs commented:
Thanks for resurrecting this, folks! The benchmarked performance improvement still looks pretty great -- can't wait to see it in production!

milleruntime commented:
> Also, I would be super in-favor with including this in Accumulo. I think there are some licensing things that we would have to figure out (JMH is GPL IIRC), but I think this would be worthwhile as a benchmark is no good if no one else ever runs it :)

I like the idea.  Any ideas how we would record benchmarks?  Have them run with the release plugin? Then publish with release notes?  

We should also discuss alternatives to JMH.  I only chose it because it was the first one you suggested.   FYI I haven't been able to get the setup() and teardown() annotations to work.

joshelser commented:
> Any ideas how we would record benchmarks? Have them run with the release plugin? Then publish with release notes?

I'm hesitant to suggest publishing them with release notes as the "norm". I'd think that we could just start a new part of the website to capture this (since they are likely only developer-related). If we feel like it is worthy of mentioning in the release notes (e.g. the numbers are not misleading or context-specific), then we can do so. Otherwise, the website can just serve as the record of "results". We could expand this out to include things like ContinuousIngest numbers too (which would be awesome -- don't have to build up a list of numbers from release notes..)

> We should also discuss alternatives to JMH. I only chose it because it was the first one you suggested. FYI I haven't been able to get the setup() and teardown() annotations to work.

Heh. I think I played around with JMH once. For the most part, I think JMH and Caliper are pretty equivalent in terms of functionality. @matthew-dailey, didn't I talk to you about this a long time ago and you had some experience with them both?

scubafuchs commented:
Here's a published benchmark example that I've found valuable in the past: https://home.apache.org/~mikemccand/lucenebench/

matthew-dailey commented:
@joshelser we did talk about this, but I only have JMH experience.  Looking through our convos, I found [this](http://www.google.com/url?q=http%3A%2F%2Fhg.openjdk.java.net%2Fcode-tools%2Fjmh%2Ffile%2Ffa510264b3f6%2Fjmh-samples%2Fsrc%2Fmain%2Fjava%2Forg%2Fopenjdk%2Fjmh%2Fsamples%2FJMHSample_11_Loops.java%23l47&sa=D&sntz=1&usg=AFQjCNF22nEbxn2-JkEhIy2LDeM3iimXpw) and [this](https://groups.google.com/forum/#!topic/mechanical-sympathy/m4opvy4xq3U) as references about JMH being good.  The gist is that JMH is made by OpenJDK developers, so they did a pretty good job designing JMH to get around things like JIT optimizations and properly measure performance.

milleruntime commented:
Aside from discussion of benchmark tests in Accumulo, any more input about the actual code changes?  I am currently running the tests again on my Ubuntu laptop to get a set of Linux system comparisons.

joshelser commented on core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java:
```diff
@@ -746,50 +746,45 @@ public boolean hasTop() {
     @Override
     public void next() throws IOException {
```
Why the consolidation of `next()` and `_next()` into just `next()`?

scubafuchs replied:
A bunch of these changes are meant to reduce the call stack size and make the jit's job easier. I don't remember if the jit has trouble with optimizing this call in particular, or whether that has changed since the original tests.

joshelser replied:
Ok -- @milleruntime this is one example of something that would be good to check is still the case. I *assume* some of what Adam tested on his own is still applicable but it would be good.

I think we can see this sort of stuff with special JVM opts, but can't point you to it off the cuff.

milleruntime replied:
I don't think this method is a problem anymore for the JIT.  There were other methods in Rfile that popped up though, I listed all the ones I found here: https://issues.apache.org/jira/browse/ACCUMULO-4621?jql=project%20%3D%20ACCUMULO

joshelser commented on core/src/main/java/org/apache/accumulo/core/iterators/ServerFilter.java:
```diff
@@ -0,0 +1,80 @@
+  private void findTop() throws IOException {
+    while (source.hasTop()) {
+      Key top = source.getTopKey();
+      if (top.isDeleted() || (accept(top, source.getTopValue()))) {
```
This condition seems strange. We want to accept deletes we see?

I'd have expected `if (!top.isDeleted() && accept(top, source.getTopValue()))`, avoiding the `accept()` call when we know the key is actually a delete.

keith-turner replied:
If `top.isDeleted()` returns true, then would `||` short circuit and not execute accept?  I think so.

joshelser replied:
That's how I read it.

keith-turner replied:
I think this code is just letting deletes pass through for compaction and not forcing users to think about deletes.

joshelser replied:
But the `break` is going to pop out of the `while`, no?

I could see:

```java
if (top.isDeleted()) {
  continue;
} else if (accept(top, source.getTopValue()) {
  break;
}
```

keith-turner replied:
I think breaking out of the while and leaving the delete on top of the source (not calling next) is the goal.  This allows the delete to pass through for compaction.

joshelser replied:
Hm. What about `IteratorUtil.setupSystemScanIterators(..)` though?

```java
  public static SortedKeyValueIterator<Key,Value> setupSystemScanIterators(SortedKeyValueIterator<Key,Value> source, Set<Column> cols, Authorizations auths,
      byte[] defaultVisibility) throws IOException {
    DeletingIterator delIter = new DeletingIterator(source, false);
    ColumnFamilySkippingIterator cfsi = new ColumnFamilySkippingIterator(delIter);
    ColumnQualifierFilter colFilter = new ColumnQualifierFilter(cfsi, cols);
    return new VisibilityFilter(colFilter, auths, defaultVisibility);
  }
```

We wouldn't even get deletes at this point, right?

scubafuchs replied:
Deletes need to pass through this filter. If we filter out a delete then we could end up un-deleting records in a partial compaction. The break happens as an indication that a top key is ready.

joshelser replied:
Gotcha. Thanks, Adam. I was thinking that the DeletingIterator would mask all deletes propagating but forgot about the partial compaction case (again)

joshelser commented on core/src/main/java/org/apache/accumulo/core/iterators/ServerSkippingIterator.java:
```diff
@@ -0,0 +1,47 @@
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.data.Value;
+
+public abstract class ServerSkippingIterator extends ServerWrappingIterator {
```
Javadoc, please.

joshelser replied:
Similarly, to the same question as to "why do we need ServerWrappingIterator and WrappingIterator", why do we need "ServerSkippingIterator" and not just "SkippingIterator"?

scubafuchs replied:
Basically, because of inheritance.

milleruntime replied:
Specifically, do you mean because the new Server* iterators directly access the protected source?  While the original WrappingIterator had a private source that was accessed through a getter and setter.

scubafuchs replied:
Yes, in the case of ServerSkippingIterator it's because we wanted to do the same thing that's done in ServerWrappingIterator. ServerWrappingIterator exists because direct access to a final source wasn't feasible in WrappingIterator.

joshelser commented on core/src/main/java/org/apache/accumulo/core/iterators/ServerWrappingIterator.java:
```diff
@@ -0,0 +1,80 @@
+
+  @Override
+  public SortedKeyValueIterator<Key,Value> deepCopy(IteratorEnvironment env) {
+    throw new UnsupportedOperationException();
```
Why the decision to throw this exception and not let the compiler tell the user (at compile time, not run time) that they need to implement this method?

scubafuchs replied:
At least at the time, I believe that a bunch of the extending classes would have implement this method the same way, so it reduces the amount of code to have a common implementation here. I think this is more of a question about whether deepCopy should really be common across all uses of the SortedKeyValueIterator interface, or whether we should have organized it differently. Interesting topic, but probably shouldn't hold up this project.

milleruntime replied:
FYI WrappingIterator currently does this for deepCopy but not init

joshelser commented on core/src/main/java/org/apache/accumulo/core/iterators/ServerWrappingIterator.java:
```diff
@@ -0,0 +1,80 @@
+
+  @Override
+  public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options, IteratorEnvironment env) throws IOException {
+    throw new UnsupportedOperationException();
```
Ditto WRT the `deepCopy()` comment

scubafuchs replied:
Same as above -- a bunch of the system iterators don't use or support the init method, so this is a common implementation.

joshelser commented on core/src/main/java/org/apache/accumulo/core/iterators/ServerWrappingIterator.java:
```diff
@@ -0,0 +1,80 @@
+ * this means that <code>getSource().seek</code> and <code>super.seek</code> no longer perform identical actions. Implementors should take note of this and if
+ * <code>seek</code> is overridden, ensure that <code>super.seek</code> is called before data is read.
+ */
+public abstract class ServerWrappingIterator implements SortedKeyValueIterator<Key,Value> {
```
How/why does this need to exist when compared to the `WrappingIterator`. Going off memory, these are very similar.

scubafuchs replied:
Two differences: the source member is exposed directly to extending classes, and there is a constructor to set the source rather than a setSource() method. This allows the source to be final which I think maybe helps with runtime optimizations?

keith-turner replied:
Another diff w/ Wrapping iterator is there are sanity checks every time `getSource()` is called.  ServerWrappingIterator does have these checks.

joshelser replied:
Ok, so essentially we're creating custom variants of things that are generally "less safe" because we're controlling what the implementations are. I can buy that -- would suggest a comment to that extent as well.

joshelser commented on core/src/main/java/org/apache/accumulo/core/iterators/system/VisibilityFilter.java:
```diff
@@ -16,63 +16,59 @@
 import org.slf4j.LoggerFactory;
 
-public class VisibilityFilter extends Filter {
+public class VisibilityFilter extends SynchronizedServerFilter {
```
This looks to me that, for every call to this Filter we're now grabbing a lock?

That sounds *really* bad to me. What's the reasoning here?

scubafuchs replied:
We used to have a separate SynchronizedIterator on scan scope that wrapped all of the system iterators and this moves that same functionality down into the VisibilityFilter. The purpose of synchronization is to prevent leakage of intermediate values through methods like getTopKey and getTopValue while another thread is calling next. This is only necessary when a configurable iterator does something multi-threaded. There are at least a couple of ways to avoid the need to make this iterator synchronized: (1) make the system iterators thread safe as exposed by the top iterator in some other way, at least from a security perspective, or (2) detect and/or disallow any use of threading in configurable iterators. Maybe you can do (1) by eliminating the possibility of leaking top keys and values that are still being evaluated by the filter? Probably needs a test to make sure we don't break that in the future.

joshelser replied:
> We used to have a separate SynchronizedIterator on scan scope that wrapped all of the system iterators and this moves that same functionality down into the VisibilityFilter.

Ok, we def need to lift your great explanation into some source code comments :)

> This is only necessary when a configurable iterator does something multi-threaded.

Ah yes. IMO, we should just tell people "don't do that" and move on, but that's an issue for another day :P

phrocker replied:
^ I second this. I think having that history embedded somewhere in comments if it is not already is incredibly useful.

milleruntime commented:
@scubafuchs Would you be able to provide feedback to Josh about your original changes?

joshelser commented:
> @scubafuchs Would you be able to provide feedback to Josh about your original changes?

Sorry, I didn't realize these were copy-paste of Adam's old changes. Let me go re-read history.

milleruntime commented:
> I'm quite astounded that the benchmarks showed these changes are faster than previous (notably, the SynchronizedFilter).

It is possible some of the changes are not improvements and the changes that are useful are outperforming the negative ones.  I don't think its sensible to go through and run performance tests on each change individually but if there are things that are obvious (like SynchronizedFilter), I can remove it to see if it makes a difference.

milleruntime commented:
Added performance benchmarks for my Ubuntu Laptop: 
[testresults.txt](https://github.com/apache/accumulo/files/945068/testresults.txt)

keith-turner commented on core/src/main/java/org/apache/accumulo/core/iterators/WrappingIterator.java:
```diff
@@ -58,21 +58,21 @@ protected void setSource(SortedKeyValueIterator<Key,Value> source) {
     if (seenSeek == false)
       throw new IllegalStateException("never been seeked");
-    return getSource().getTopKey();
+    return source.getTopKey();
```
This could possibly break iterators that extend WrappingIterator and override getSource().  I don't think we should make this change.

Maybe it would be better to deprecate WrappingIterator and introduce another version for users that more efficient.  I don't think we have a lot of wiggle room to make this one more efficient without possibly breaking user iterators.

scubafuchs replied:
In Sqrrl we use a LightWrappingIterator that does this. I'd be in favor of deprecating the WrappingIterator, but we might want to couple that with the bigger iterator rewrite project (ACCUMULO-3751).

keith-turner replied:
Is LightWrappingIterator the same as ServerWrappingIterator in this PR?

scubafuchs replied:
No, LightWrappingIterator is a user-level iterator, so it supports a no-argument constructor and non-final source to be compatible with the initialization via init method. It basically just removes some check, doesn't call getSource(), and exposes the source field as protected instead of private.

keith-turner replied:
Interesting, the init method prevents making the source final.  Didn't see that one coming.

keith-turner replied:
I experimented with making the VersioningIterator and Filter not call WrappingIterator.getSource() multiple times in their loops.  I saw a noticeable difference, like 5% increase for the VersioningIterator.  I change the code to call getSource() once before the loop and store the source in a local var.

milleruntime commented:
Reverted changes to RFile (they didn't seem very useful) and reverted changes to WrappingIterator since they would break user iterators in 180b9f6
Tests still show the same level of improvements

milleruntime commented on server/tserver/src/main/java/org/apache/accumulo/tserver/InMemoryMap.java:
```diff
@@ -683,13 +683,6 @@ public void setInterruptFlag(AtomicBoolean flag) {
-        throw new IllegalStateException("Memory iterator is closed");
-      return super.getSource();
-    }
-
```
Without the WrappingIterator changes, I don't think we gain anything from this change... we just lose the closed check.

milleruntime replied:
I am going to experiment to see how much it affects performance and if there is a compromise where we only have the sanity check at the start of loops.

milleruntime replied:
I actually saw performance improvements when I added this method back... which I don't understand at all.

milleruntime replied:
It could just be a margin of error or the JMH test I wrote: https://github.com/milleruntime/jmh-test/blob/6cd94569bbecda42181b6b7d7101e0770ef00d7f/src/main/java/org/sample/InMemoryMapBenchmark.java

keith-turner replied:
I did a bit of research into the implications of removing this sanity check.  The case where the check would have triggered is when scan iterators are used after Accumulo has finished using the iterator stack.  If that's what we want to check for, this is probably not the best place to do it.

milleruntime commented:
My attempt at proper java doc for the new server iterators in 12a0bcf

milleruntime commented:
I think we have a good bunch of improvements here, that I would like to merge today.  If there is any more feedback please let me know.  I can open separate tickets for the other ideas that we discussed here (incorporating benchmarks into main repo, deprecating WrappingIterator for a Light version).

milleruntime commented:
Updated test results on Linux.  Dropping the lowest and highest gains, we average at about an 80% gain across the 8 tests. 
[JMH Benchmarks](https://github.com/milleruntime/jmh-test/blob/master/src/main/java/org/sample/MyBenchmark.java)
[testresults.txt](https://github.com/apache/accumulo/files/962199/testresults.txt)

scubafuchs commented:
Improvements definitely look good. What's up with the benchmarks showing the *1000 tests being about 200x slower than the *10 tests when they are only scanning over 100x the data? Any insights?

milleruntime commented:
> What's up with the benchmarks showing the *1000 tests being about 200x slower than the *10 tests when they are only scanning over 100x the data? Any insights?

The tests names are deceiving since the numbers are formatted in HEX.  So its more like 16 keys and 4096.

scubafuchs commented:
Ah, that would explain it!

milleruntime commented:
For the skeptics out there, I understand that when you take into consideration all the other factors of a cluster, low level system iterators only make up a small percentage of system performance considerations.  That being said, Accumulo developers have the most control over lower level code such as these system iterators so I think any improvements we can make are worthwhile.

joshelser commented:
> I understand that when you take into consideration all the other factors of a cluster, low level system iterators only make up a small percentage of system performance considerations

Good on you for following through here. I appreciate the testing efforts you put forward.
